### PR TITLE
Support of gdn kernel from tpu-inference

### PR DESCRIPTION
**Pull request by khatwanimohit**
# Description

- This PR is inspired from @NicoGrande's PR #3917. ( Thank you @NicoGrande, we miss you! 😊 )
- GDN Integration and Sharding Support: Integrated Gated Delta Net (GDN) logic from  nicogrande/add-gdn-support .
- GDN Expert Replication: Added experimental  self._gdn_replicate_expert  option (triggered via the  MAXTEXT_GDN_REPLICATE_EXPERT  environment variable) to control whether  attn_head  is set to ShardingAxisName.MODEL  or  ShardingAxisName.ATTN_HEAD.
- Unit Test Fixes: Resolved initialization errors in  attention_test.py  by ensuring  self.mesh  is properly initialized
- Adds support for profiling in vllm_decode.py

You can also provide a comma-separated list. If you don't want to close a bug but
simply to reference it, use BUGS, e.g.: 
BUGS: b/517158881

*Notice 1:* Once all tests pass, the "pull ready" label will automatically be assigned.
This label is used for administrative purposes. Please do not add it manually.

*Notice 2:* For external contributions, our settings currently require an approval from a MaxText maintainer to trigger CI tests.

# Tests

```
NEW_MODEL_DESIGN=1 \
    PYTHONPATH=/home/mohitkhatwani_google_com/workspace/maxtext/src \
    MAXTEXT_GDN_REPLICATE_EXPERT=true \
    /home/mohitkhatwani_google_com/workspace/max_venv/bin/python3 \
      -m maxtext.inference.vllm_decode src/maxtext/configs/base.yml \
      base_output_directory=gs://runner-maxtext-logs \
      run_name=mohit-qwen3.5-maxtext-bench-$RANDOM \
      model_name=qwen3.5-35b-a3b \
      tokenizer_path=Qwen/Qwen3.5-35B-A3B \
      vllm_hf_overrides='{architectures: ["MaxTextForCausalLM"]}' \
      load_parameters_path=gs://maxtext-model-checkpoints/qwen3.5-35b-a3b/unscanned/0/items \
      ici_tensor_parallelism=4 \
      hbm_utilization_vllm=0.5 \
      prompt="Tell me three fun facts about Buenos Aires." \
      decode_sampling_temperature=0.0 decode_sampling_nucleus_p=1.0 decode_sampling_top_k=0.0 \
      pure_nnx_decoder=True use_chat_template=True max_num_seqs=1 \
      enable_dp_attention=False prefuse_moe_weights=True max_target_length=64 debug_sharding=False \
      scan_layers=false profiler=xplane
 ```
 
 Profile: https://xprof.corp.google.com/trace_viewer/mohitkhatwani-11546393038070241787?view_start=71184.720&view_end=71204.219
 
 Logs: https://paste.googleplex.com/5464600022745088
 
 Decode performance: 25ms

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/models/qwen3.py [hunk 1]
```diff
@@ -381,13 +385,25 @@ class Qwen3NextGatedDeltaNet(nnx.Module):
   2. output = Linear_out(y)
   """
 
-  def __init__(self, config: Config, dtype: DType = jnp.float32, model_mode: str = MODEL_MODE_TRAIN, *, rngs: nnx.Rngs):
+  def __init__(
+      self,
+      config: Config,
+      mesh=None,
```
Comment (xuefgu) [inline src/maxtext/models/qwen3.py:457]: The constructor signature makes `mesh` optional, but the forward pass unconditionally uses `self.mesh` in `jax.shard_map` (via `shard_mapped_delta_rule`) for the standard (non-paged) path. If `mesh` is not provided, this will crash during the forward pass.

We should either make `mesh` a required argument in `__init__` (and update docstrings), or add a fallback path in `__call__` that bypasses `shard_map` when `mesh` is `None` (which is useful for single-device CPU/GPU debugging).

File: src/maxtext/inference/vllm_decode.py [hunk 2]
```diff
@@ -52,6 +53,11 @@
 
 adapter.register()
 
+# Force uses_mrope to False to disable 3D multimodal position IDs in text-only runs.
+from vllm.config import ModelConfig
+
+ModelConfig.uses_mrope = property(lambda _: False)
```
Comment (xuefgu) [inline src/maxtext/inference/vllm_decode.py:60]: Globally monkey-patching `ModelConfig.uses_mrope` on the class level is fragile. Is there a way to configure this at the engine initialization level (e.g., via `vllm_args` or `ModelConfig` instance arguments) instead of patching the class? If this is a temporary workaround for a vLLM limitation, please add a TODO linking to the tracking issue.

File: src/maxtext/integration/vllm/maxtext_vllm_adapter/adapter.py [hunk 3]
```diff
@@ -324,3 +341,159 @@ def load_weights(self, rng_key: jax.Array) -> None:
...
+      value_dim = head_v_dim * num_v_heads
+      conv_dim = key_dim * 2 + value_dim
+
+      conv_state_shape = (conv_kernel_size - 1, conv_dim)
+      recurrent_state_shape = (num_v_heads, head_k_dim, head_v_dim)
+
+      mamba_shapes = (conv_state_shape, recurrent_state_shape)
+
+      torch_dtype = torch.bfloat16
+      if str(cfg.dtype) == "float32":
```
Comment (xuefgu) [inline src/maxtext/integration/vllm/maxtext_vllm_adapter/adapter.py:423]: Can the dtype be `float16`? If so, we want to handle that explicitly.

File: src/maxtext/integration/vllm/maxtext_vllm_adapter/adapter.py [hunk 4]
```diff
@@ -324,3 +341,159 @@ def load_weights(self, rng_key: jax.Array) -> None:
...
+      num_layers = cfg.base_num_decoder_layers
+      num_attn = num_layers // interval
+      num_mamba = num_layers - num_attn
+
+      min_count = min(num_attn, num_mamba)
+      max_count = max(num_attn, num_mamba)
+      if max_count < min_count * 1.5:
+        group_size = max_count
+      else:
+        group_size = min_count
```
Comment (xuefgu) [inline src/maxtext/integration/vllm/maxtext_vllm_adapter/adapter.py]: The logic to `calculate group_size` and the 1.5 multiplier threshold is not immediately obvious.

Could you add a brief comment explaining the design choice behind this grouping logic and how it relates to uniform page size allocation in `tpu_inference`? This will help future maintainers understand the math.

**Review by xuefgu — CHANGES_REQUESTED** [review]

File: src/maxtext/models/qwen3.py (hunk 1 above)
Comment (khatwanimohit) [reply src/maxtext/models/qwen3.py:457]: Thanks for catching this. I've updated the forward pass to include a fallback path. If  self.mesh  is  None  (e.g., during single-device CPU/GPU debugging), we bypass  jax.shard_map  completely and directly call  jax_chunk_gated_delta_rule

File: src/maxtext/inference/vllm_decode.py (hunk 2 above)
Comment (khatwanimohit) [reply src/maxtext/inference/vllm_decode.py:60]: Class-level monkey-patching is unfortunately necessary here because  ModelConfig.uses_mrope  is evaluated internally inside the  LLM(...)  constructor when the engine initializes (which happens before the  LLM  instance is returned, making instance-level configuration impossible). I've added a TODO linking to the tracking issue to see if we can expose this configuration parameter in upstream vLLM.

File: src/maxtext/integration/vllm/maxtext_vllm_adapter/adapter.py (hunk 3 above)
Comment (khatwanimohit) [reply src/maxtext/integration/vllm/maxtext_vllm_adapter/adapter.py:423]: Good point. I've updated the mapping logic to handle  float16  explicitly so that  torch_dtype  maps correctly to  torch.float16 , and configured the  dtype_size  to be set to 4 bytes for  float32  and 2 bytes for 16-bit float types.

File: src/maxtext/integration/vllm/maxtext_vllm_adapter/adapter.py (hunk 4 above)
Comment (khatwanimohit) [reply src/maxtext/integration/vllm/maxtext_vllm_adapter/adapter.py]: I've added comments explaining the design decisions behind the grouping math and the  1.5  multiplier threshold.

File: src/maxtext/integration/vllm/maxtext_vllm_adapter/adapter.py [hunk 5]
```diff
@@ -324,3 +341,168 @@ def load_weights(self, rng_key: jax.Array) -> None:
...
+      # To allocate memory uniformly for a hybrid model's KV/recurrent cache page table,
+      # we group layers together. The uniform page size must support both attention and
+      # mamba layers.
+      # If the ratio of attention to mamba layers is relatively balanced (less than 1.5x),
+      # we use the larger count as the group size to minimize the total number of groups.
+      # If they are highly imbalanced (>= 1.5x), we group by the smaller count to prevent
+      # the page size from being inflated by excessive padding for the minority layer type.
+      min_count = min(num_attn, num_mamba)
+      max_count = max(num_attn, num_mamba)
+      if max_count < min_count * 1.5:
```
Comment (xuefgu) [inline src/maxtext/integration/vllm/maxtext_vllm_adapter/adapter.py]: Can we make this 1.5 a constant with comment?

**Review by xuefgu — APPROVED** [review]
A nit but otherwise LGTM. Thanks for the PR!

**Review by A9isha — APPROVED** [review]
Thanks Mohit!
